src/lib/storage/abstract_device_factory.rb

```diff
@@ -0,0 +1,317 @@
+#!/usr/bin/env ruby
+#
+# encoding: utf-8
+
+# Copyright (c) [2015] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "storage"
+require "yaml"
+
+module Yast
+  module Storage
+    #
+    # Abstract factory class to generate device trees and similar objects with
+    # a tree structure from YaML. The FakeDeviceFactory is one example subclass.
+    #
+    # This class uses introspection and duck-typing with a number of predefined
+    # methods that a subclass is required to implement.
+    #
+    # Subclasses are required to implement a create_xy() method for each
+    # factory product 'xy' the factory is able to create and some methods to
```

Comment:
Actually, no:

"Subclasses are required to implement a create_xy() method for each factory product 'xy' the factory is able to create and some methods to support some basic sanity checks of the generated device tree"

Subclasses are required to implement

-   a create_xy() method for each factory product 'xy' the factory is able to create

and

- some methods to support some basic sanity checks